README
NAME
    AnyEvent::Connection - Base class for tcp connectful clients

SYNOPSIS
        package MyTCPClient;
        use base 'AnyEvent::Connection';

        package main;
        my $client = MyTCPClient->new(
        host => 'localhost',
        port => 12345,
        );
        $client->reg_cb(
        connected => sub {
            my ($client,$connection,$host,$port) = @_;
            # ...
            $client->after(
            $interval, sub {
                # Called after interval, if connection still alive
            }
            );
        }
        connfail = sub {
            my ($client,$reason) = @_;
            # ...
        },
        disconnect => sub {
            my ($client,$reason) = @_;
        },
        error => sub {
            my ($client,$error) = @_;
            # Called in error conditions for callbackless methods
        },
        );
        $client->connect;

EVENTS
    connected ($connobject, $host, $port)
        Called when client get connected.

    connfail
        Called, when client fails to connect

    disconnect
        Called whenever client disconnects

    error
        Called in error conditions for callbackless methods (for ex: when
        calling push_write on non-connected client)

OPTIONS
    host
        Host to connect to

    port
        Port to connect to

    timeout [ = 3 ]
        Connect/read/write timeout in seconds

    reconnect [ = 1 ]
        If true, automatically reconnect after disconnect/connfail after
        delay $reconnect seconds

    rawcon [ = AnyEvent::Connection::Raw ]
        Class that implements low-level connection

OPERATION METHODS
    new Cleates connection object (see OPTIONS)

    connect
        Begin connection

    disconnect ($reason)
        Close current connection. reason is optional

    reconnect
        Close current connection and establish a new one

    after($interval, $cb->())
        Helper method. AE::timer(after), associated with current connection

        Will be destroyed if connection is destroyed, so no timer invocation
        after connection destruction.

    periodic($interval, $cb->())
        Helper method. AE::timer(periodic), associated with current
        connection

        Will be destroyed if connection is destroyed, so no timer invocation
        after connection destruction.

    periodic_stop()
        If called within periodic callback, periodic will be stopped.

        my $count;
        $client->periodic(1,sub {
            $client->periodic_stop if ++$count > 10;
        });

        # callback will be called only 10 times;

    destroy
        Close connection, destroy all associated objects and timers, clean
        self

CONNECT METHODS
    When connected, there are some methods, that proxied to raw connection
    or to AE::Handle

    push_write
        See AE::Handle::push_write

    push_read
        See AE::Handle::push_read

    unshift_read
        See AE::Handle::unshift_read

    say Same as push_write + newline

    reply
        Same as push_write + newline

    For next methods there is a feature. Callback will be called in any way,
    either by successful processing or by error or object destruction

    recv($bytes, %args, cb => $cb->())
        Similar to

        $fh->push_read(chunk => $bytes, $cb->());

    command($data, %args, cb => $cb->());
        Similar to

        $fh->push_write($data);
        $fh->push_read(line => $cb->());

AUTHOR
    Mons Anderson, "<mons at cpan.org>"

BUGS
    Please report any bugs or feature requests to "bug-anyevent-connection
    at rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=AnyEvent-Connection>. I
    will be notified, and then you'll automatically be notified of progress
    on your bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc AnyEvent::Connection

    You can also look for information at:

    *   RT: CPAN's request tracker

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=AnyEvent-Connection>

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/AnyEvent-Connection>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/AnyEvent-Connection>

    *   Search CPAN

        <http://search.cpan.org/dist/AnyEvent-Connection/>

ACKNOWLEDGEMENTS
COPYRIGHT & LICENSE
    Copyright 2009 Mons Anderson, all rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for AnyEvent-Connection

0.06_01 2010-12-03
        Fix operation on destroyed connection

0.06    2010-11-16
        Fix syntax for bleadperl (5.13+)

0.05    2010-02-18
        Remove (hide) development dependencies
        Move author tests to xt/

0.04    2009-11-10
        Fix deps

0.03    2009-11-10
        Fix deps

0.02    2009-11-09
        Fixes, tests

0.01    2009-09-24
        First version



lib/AnyEvent/cb.pm
package AnyEvent::cb;

use strict;
use warnings;
use Scalar::Util qw(weaken);

our %DEF;

BEGIN {
	if ($ENV{CB_DEBUG}) {
		*DEBUG = sub () { 1 };
	} else {
		*DEBUG = sub () { 0 };
	}
}
BEGIN {
	sub sb(&);
	sub cb(&;@);
	if (DEBUG) {
		eval { require Sub::Identify; Sub::Identify->import('sub_fullname'); 1 } or *sub_fullname = sub { return };
		eval { require Devel::Refcount; Devel::Refcount->import('refcount'); 1 } or *refcount = sub { 1 };
		eval { require Devel::FindRef; *findref = \&Devel::FindRef::track;   1 } or *findref  = sub { "No Devel::FindRef installed\n" };
		*sb = sub (&) {
			$DEF{int $_[0]} = [ join(' ',(caller())[1,2]), $_[0] ];weaken($DEF{int $_[0]}[1]);
			#local $_ = int $_[0];
			#warn "create: $_ defined $DEF{$_}[0]\n";
			return bless shift,'__callback__';
		};
		*cb = sub (&;@) {
			$DEF{int $_[0]} = [ join(' ',(caller())[1,2]), $_[0] ];weaken($DEF{int $_[0]}[1]);
			return cb => bless( shift,'__callback__'), @_;
		};
		*__callback__::DESTROY = sub {
			#local $_ = int $_[0];
			#my $name = sub_fullname($DEF{$_}[1]);
			#warn "destroy: $_ ".($name ? $name : 'ANON')." defined $DEF{$_}[0]\n";#.findref($DEF{$_}[1]);
			delete($DEF{int $_[0]});
		};
		*COUNT = sub {
			for (keys %DEF) {
				$DEF{$_}[1] or next;
				my $name = sub_fullname($DEF{$_}[1]);
				warn "Not destroyed: $_ ".($name ? $name : 'ANON')." (".refcount($DEF{$_}[1]).") defined $DEF{$_}[0]\n";#.findref($DEF{$_}[1]);
			}
		};
	} else {
		*sub_fullname = sub {};
		*refcount = sub { 1 };
		*findref = sub {};
		*sb = sub (&) { $_[0] };
		*cb = sub (&;@) { cb => shift, @_ };
		*COUNT = sub {};
	}
}


sub import {
	no strict 'refs';
	for (qw(sb cb)) {
		*{ caller().'::'.$_ } = \&$_;
	}
}

END {
	COUNT() if DEBUG;
}

1;


lib/AnyEvent/Connection.pm
package AnyEvent::Connection;

use common::sense 2;m{
use strict;
use warnings;
};
use Object::Event 1.21;
use base 'Object::Event';

use AnyEvent 5;
use AnyEvent::Socket;

use Carp;

use Scalar::Util qw(weaken);
use AnyEvent::Connection::Raw;
use AnyEvent::Connection::Util;
# @rewrite s/^# //; # Development hacks, see L<Devel::Rewrite>
# use Devel::Leak::Cb;

=head1 NAME

AnyEvent::Connection - Base class for tcp connectful clients

=cut

our $VERSION = '0.06_01'; $VERSION = eval($VERSION);

=head1 SYNOPSIS

    package MyTCPClient;
    use base 'AnyEvent::Connection';

    package main;
    my $client = MyTCPClient->new(
        host => 'localhost',
        port => 12345,
    );
    $client->reg_cb(
        connected => sub {
            my ($client,$connection,$host,$port) = @_;
            # ...
            $client->after(
                $interval, sub {
                    # Called after interval, if connection still alive
                }
            );
        }
        connfail = sub {
            my ($client,$reason) = @_;
            # ...
        },
        disconnect => sub {
            my ($client,$reason) = @_;
        },
        error => sub {
            my ($client,$error) = @_;
            # Called in error conditions for callbackless methods
        },
    );
    $client->connect;

=head1 EVENTS

=over 4

=item connected ($connobject, $host, $port)

Called when client get connected.

=item connfail

Called, when client fails to connect

=item disconnect

Called whenever client disconnects

=item error

Called in error conditions for callbackless methods (for ex: when calling push_write on non-connected client)

=back

=head1 OPTIONS

=over 4

=item host

Host to connect to

=item port

Port to connect to

=item timeout [ = 3 ]

Connect/read/write timeout in seconds

=item reconnect [ = 1 ]

If true, automatically reconnect after disconnect/connfail after delay $reconnect seconds

=item rawcon [ = AnyEvent::Connection::Raw ]

Class that implements low-level connection

=back

=head1 OPERATION METHODS

=over 4

=item new

Cleates connection object (see OPTIONS)

=item connect

Begin connection

=item disconnect ($reason)

Close current connection. reason is optional

=item reconnect


Close current connection and establish a new one

=item after($interval, $cb->())

Helper method. AE::timer(after), associated with current connection

Will be destroyed if connection is destroyed, so no timer invocation after connection destruction.

=item periodic($interval, $cb->())

Helper method. AE::timer(periodic), associated with current connection

Will be destroyed if connection is destroyed, so no timer invocation after connection destruction.

=item periodic_stop()

If called within periodic callback, periodic will be stopped.

    my $count;
    $client->periodic(1,sub {
        $client->periodic_stop if ++$count > 10;
    });
    
    # callback will be called only 10 times;

=item destroy

Close connection, destroy all associated objects and timers, clean self

=back

=head1 CONNECT METHODS

When connected, there are some methods, that proxied to raw connection or to AE::Handle


=over 4

=item push_write

See AE::Handle::push_write

=item push_read

See AE::Handle::push_read

=item unshift_read

See AE::Handle::unshift_read

=item say

Same as push_write + newline

=item reply

Same as push_write + newline

=back

For next methods there is a feature.
Callback will be called in any way, either by successful processing or by error or object destruction

=over 4

=item recv($bytes, %args, cb => $cb->())

Similar to

    $fh->push_read(chunk => $bytes, $cb->());

=item command($data, %args, cb => $cb->());

Similar to

    $fh->push_write($data);
    $fh->push_read(line => $cb->());

=back

=cut

sub new {
	my $self = shift->SUPER::new(@_);
	$self->init(@_);
	return $self;
}

sub init {
	my $self = shift;
	$self->{debug}   ||= 0;
	$self->{connected} = 0;
	$self->{connecting} = 0;
	$self->{reconnect} = 1 unless defined $self->{reconnect};
	$self->{timeout} ||= 3;
	$self->{timers}    = {};
	$self->{rawcon}  ||= 'AnyEvent::Connection::Raw';
	#warn "Init $self";
}

#sub connected {
#	warn "Connected";
#	shift->event(connected => ());
#}

sub connect {
	my $self = shift;
	$self->{connecting} and return;
	$self->{connecting} = 1;
	weaken $self;
	croak "Only client can connect but have $self->{type}" if $self->{type} and $self->{type} ne 'client';
	$self->{type} = 'client';
	
	warn "Connecting to $self->{host}:$self->{port}..." if $self->{debug};
	# @rewrite s/sub {/cb connect {/;
	$self->{_}{con}{cb} = sub {
		pop;
		delete $self->{_}{con};
			if (my $fh = shift) {
				warn "Connected @_" if $self->{debug};
				$self->{con} = $self->{rawcon}->new(
					fh      => $fh,
					timeout => $self->{timeout},
					debug   => $self->{debug},
				);
				$self->{con}->reg_cb(
					disconnect => sub {
						warn "Disconnected $self->{host}:$self->{port} @_";# if $self->{debug};
						$self->disconnect(@_);
						$self->_reconnect_after();
					},
				);
				$self->{connected} = 1;
				#warn "Send connected event";
				$self->event(connected => $self->{con}, @_);
			} else {
				warn "Not connected $self->{host}:$self->{port}: $!" if $self->{debug};
				$self->event(connfail => "$!");
				$self->_reconnect_after();
			}
	};
	$self->{_}{con}{pre} = sub { $self->{timeout} };
	$self->{_}{con}{grd} =
		AnyEvent::Socket::tcp_connect
			$self->{host}, $self->{port},
			$self->{_}{con}{cb}, $self->{_}{con}{pre}
	;
}

sub accept {
	croak "Not implemented yet";
}


sub _reconnect_after {
	weaken( my $self = shift );
	$self->{reconnect} or return $self->{connecting} = 0;
	$self->{timers}{reconnect} = AnyEvent->timer(
		after => $self->{reconnect},
		cb => sub {
			$self or return;
			delete $self->{timers}{reconnect};
			$self->{connecting} = 0;
			$self->connect;
		}
	);
}

sub periodic_stop;
sub periodic {
	weaken( my $self = shift );
	my $interval = shift;
	my $cb = shift;
	#warn "Create periodic $interval";
	$self->{timers}{int $cb} = AnyEvent->timer(
		after => $interval,
		interval => $interval,
		cb => sub {
			local *periodic_stop = sub {
				warn "Stopping periodic ".int $cb;
				delete $self->{timers}{int $cb}; undef $cb
			};
			$self or return;
			$cb->();
		},
	);
	defined wantarray and return AnyEvent::Util::guard(sub {
		delete $self->{timers}{int $cb};
		undef $cb;
	});
	return;
}

sub after {
	weaken( my $self = shift );
	my $interval = shift;
	my $cb = shift;
	#warn "Create after $interval";
	$self->{timers}{int $cb} = AnyEvent->timer(
		after => $interval,
		cb => sub {
			$self or return;
			delete $self->{timers}{int $cb};
			$cb->();
			undef $cb;
		},
	);
	defined wantarray and return AnyEvent::Util::guard(sub {
		delete $self->{timers}{int $cb};
		undef $cb;
	});
	return;
}

sub reconnect {
	my $self = shift;
	$self->disconnect;
	$self->connect;
}

sub disconnect {
	my $self = shift;
	#$self->{con} or return;
	warn "Disconnecting $self->{connected} || $self->{connecting} || $self->{reconnect} by @{[ (caller)[1,2] ]}";
	ref $self->{con} eq 'HASH' and warn dumper($self->{con});
	$self->{con} and eval{ $self->{con}->close; };
	warn if $@;
	delete $self->{con};
	my $wascon = $self->{connected};# || $self->{connecting};
	warn "disconnect with wascon=$wascon $self->{connected}/$self->{connecting}";
	$self->{connected}  = 0;
	$self->{connecting} = 0;
	#$self->{reconnect}  = 0;
	delete $self->{timers}{reconnect};
	$self->event('disconnect',@_) if $wascon;
	return;
}

sub AnyEvent::Connection::destroyed::AUTOLOAD {}

sub destroy {
	my ($self) = @_;
	$self->DESTROY;
	bless $self, "AnyEvent::Connection::destroyed";
}

sub DESTROY {
	my $self = shift;
	warn "(".int($self).") Destroying AE::CNN" if $self->{debug};
	$self->disconnect;
	%$self = ();
}

BEGIN {
	no strict 'refs';
	for my $m (qw(push_write push_read unshift_read say reply recv command want_command)) {
		*$m = sub {
			my $self = shift;
			$self->{connected} or return $self->event( error => "Not connected for $m" );
			$self->{con}->$m(@_);
		};
	}
}

=head1 AUTHOR

Mons Anderson, C<< <mons at cpan.org> >>

=head1 COPYRIGHT & LICENSE

Copyright 2009 Mons Anderson, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


=cut

1; # End of AnyEvent::Connection


tmp/test-pmq.pl
#!/usr/bin/env perl -w

package main;
use strict;
use lib::abs '../lib', '../../Queue-PQ/lib';
use AnyEvent::Impl::Perl;
#use AnyEvent::Impl::EV;
use AnyEvent::Connection;
use Time::HiRes qw(time);
use Devel::FindRef;sub findref;*findref = \&Devel::FindRef::track;
use Scalar::Util qw(weaken);
use Devel::Refcount qw( refcount );
use R::Dump;
use AnyEvent::cb;

{
	my $cv = AnyEvent->condvar;
	$cv->begin(sub {$cv->send});
	my $run;
	$SIG{INT} = sub { $cv->send; };
	$0 = "test-pmq - connect";
for (1..1) {
	$cv->begin;
	my $client;
	my $see =
	$client = AnyEvent::Connection->new(
		timeout => 1,
		host => 'localhost',
		#port => $port,
		#port => '11311',
		port => '12321',
	);
	weaken($see);
	$client->reg_cb(
		error => sub {
			#$client;
			warn "Error cb @_";
		},
		connected => sb {
			warn "Connected cb";
=for rem
			$client->push_write("stats\r\n");
			$client->push_read(line => sb {
				warn "Got @_";
				undef $client;
				$cv->end;
			} 'read');
			return;
=cut
			my $count = 0;
			$client->after(0.5,sb {
				warn "Stopping";
				$client->disconnect;
				undef $client;
				$cv->end;
			});
			$client->periodic(0.1,sb {
				warn "Periodic";
				$client->periodic_stop if ++$count > 3;
			});
			return;
		},
		connfail => sb {
			warn "Connfail cb @_";
			return;
			$client->{_}{rtimer} = AnyEvent->timer(
				after=> 0.1,
				cb {
					delete $client->{_}{rtimer};
					warn "Reconnecting";
					$client->connect;
				}
			);
		},
		disconnect => sub {
			warn "Disonnected cb";
		},
	);
	$client->connect;
}
	$cv->end;
	$cv->recv;
	#undef $client;
	warn "Ending...";
	#$see and warn findref $see;
	exit 0;

}

END {
	warn "main::end";
}

__END__




#my $inverval = 0.001;
my $inverval = 0.01;
my $rate = 10;
#my $rate = 3;
my $reload = 1;
#my @queues = qw(test1 test2 test3);
my @queues = qw(test1);
my @range = 1..30;
#my @range = 1..1;
#our %history;
our %taken;
our %seen;
our $port = 12345;
our ($server,$loader,$client);

sub mem () {
	my $ps = (split /\n/,`ps auxwp $$`)[1];
	my ($mem) = $ps =~ /\w+\s+\d+\s+\S+\s+\S+\s+(\d+)\s+(\d+)/;
	return $mem;
}
our $cmem = 0;
sub measure ($) {
	my $op = shift;
	my $mem = mem;
	my $delta = $mem - $cmem;
	if ($delta != 0) {
		$cmem = $mem;
		warn sprintf "%s: %+d\n",$op,$delta;
		#warn Dump $cl;
	}
}

unless ($server = fork) {
	exit 0;
	my $cv = AnyEvent->condvar;
	$SIG{INT} = sub { $cv->send };
	$0 = "test-pmq - server";
	my $server = AnyEvent::Queue::Server::PMQ->new(
			port         => $port,
			devel        => 0,
		#debug_proto  => 1,
		#debug_engine => 1,
	);
	$server->start;
	$cv->recv;
	exit 0;
}

unless ($loader = fork) {
	exit 0;
	my $cv = AnyEvent->condvar;
	$SIG{INT} = sub { $cv->send };
	$cv->begin(sub { $cv->send });
	$0 = "test-pmq - loader";
	my $c = AnyEvent::Queue::Client::PMQ->new(
		servers => ['localhost:'.$port],
		#sync => 1,
		reconnect => 0,
		debug => 0,
	);
	$cv->begin;
	$c->connect(cb {
		warn "Loader client connected: @_";
		for my $dst (@queues) {
			for my $id (@range) {
				$cv->begin;
				$c->put(
					dst => $dst,
					id  => $id,
					data => { x => 'x'x10 },
					cb => sub {
						#push @{$history{$id}},"insert: $! ".( @_ > 1 ? "$_[1]" :'').( $taken{$id} ? ' +taken' : '' );
						shift or $!{EEXIST} or warn "create $dst.$id failed: @_";
						$cv->end;
					},
				);
			}
		}
		undef $c;
		$cv->end;
	});
	$cv->end;
	$cv->recv;
	exit 0;
}

#if (0) {
unless ($client = fork) {
	my $cv = AnyEvent->condvar;
	$SIG{INT} = sub { $cv->send };
	$0 = "test-pmq - connect";
	my $client = Connector->new(
		host => 'localhost',
		#port => $port,
		port => '11311',
	);
	$client->reg_cb(
		connected => sub {
			warn "Connected cb";
		},
		disconnected => sub {
			warn "Disonnected cb";
		},
	);
	$client->connect;
=for rem
	my $con; $con = sub {
		warn "Connecting";
		measure('connect begin');
		my $g;$g = AnyEvent::Socket::tcp_connect 'localhost', $port, sub {
			pop;
			undef $g;
			warn "Connected @_";
			my $fh = shift or warn("$!"),return @_ = ();
			my $cnn = AnyEvent::Queue::Conn->new(
				fh => $fh,
			);
			
			@_ = ();
			measure('connected');
			after { $con->() } 1;
		}, sub { 0.9 };
		measure('connect end');
	};
	$con->();
=cut
	$cv->recv;
	exit 0;
}

if (0) {
#unless ($client = fork) {
	my $cv = AnyEvent->condvar;
	$SIG{INT} = sub { $cv->send };
	$0 = "test-pmq - client";
	measure('start');
	my $c2 = AnyEvent::Queue::Client::PMQ->new(
		servers => ['localhost:12345'],
		reconnect => 1,
		timeout => 1,
	);
	measure('client');


	our %watchers;
	our %see;

	$c2->connect(cb {
		warn "Client connected @_";
		measure('connect');
		my $statscb;$statscb = sub {
			my $stats = shift;
			#warn("Got stats $stats->{queue}{test1}{taken} $stats->{queue}{test2}{taken} $stats->{queue}{test3}{taken}");
			measure('stats');
			$c2->queues(cb {
				my $q = shift or return warn("queues failed: @_");
				measure('queues');
				warn("Got stats ".join(' ',map { $stats->{queue}{$_}{taken} } @$q ));
				#warn "Source have queues: [@$q]";
				for my $sv (@$q) {
					my $key;
					if( $watchers{$sv} ) {
						#warn "Watcher: $watchers{$sv}";
						#warn "Watcher: taken @{[ %{$watchers{$sv}{taken}} ]}";
						#warn "Watcher have taken: $watchers{$sv}{taken}{count} ".$watchers{$sv}->taken_keys;
						if ($watchers{$sv}{taken}{count} == 0) {
							warn Dump $c2->{taken};
							#$cv->send if %{ $c2->{taken} };
						}
						#warn Dump $watchers{$sv};
						#print Devel::FindRef::track $watchers{$sv};
						#$cv->send;
						#weaken($see{ $key = int $watchers{$sv} } = $watchers{$sv});
					};
					weaken( my $wx = $watchers{$sv} );
					warn "Restart with taken: ".Dump $c2->{taken};
					measure('begin create watcher');
					$watchers{$sv} = $c2->watcher(
						src      => $sv,
						prefetch => $rate,
						#rate     => $rate,
						job => sub {
							my $w = shift;
							my ($job,$err) = @_;
							measure('job');
							if ($job) {
								$seen{$job->{id}}++;
								if ($taken{$sv}{$job->{id}}++) {
									warn "$sv.$job->{id} already taken but have one more\n";
									#$cv->send;
								}
								#warn "$w ++$job->{id}";
								#return;
								$see{int $w}++;
								#after {
										#warn "End timer $w / ".refcount($w);
										#$w->requeue(job => $job);return;
										$w->requeue(job => $job, cb => sub {
											#shift or warn "@_";
											delete $taken{$sv}{$job->{id}};
										});
										$see{int $w}--;
								#} 0.1;
							} else {
								warn "WTF @_?";
							}
						},
						nomore => sub {
							#warn("No more items for $sv");
						},
					);
					measure('watcher');
					after {
						if ($wx) {
							warn "Not cleaned watcher ".eval{ refcount($wx) };
							warn "See: ".$see{int $wx};
							print findref $wx;
							#$cv->send;
							die "FTW";
						} else {
							warn "Cleaned";
						}
						measure('timer end');
					} 3;
					measure('timer');
					#warn "(".int($watchers{$sv})." : ".refcount($watchers{$sv}).") New watcher";
					#print #Devel::FindRef::track $see{$key} if $key and $see{$key};
				}
			});
			measure('call queues');
			my $t;$t = AnyEvent->timer(
				cb {
					undef $t;
					measure('invoke fullstats');
					$c2->fullstats(cb => $statscb)
				}
				after => $reload,
			);
			measure('delay fullstats');
		};
		$c2->fullstats(cb => $statscb);return;
		my $sv = $queues[0];
		my $neww = sub {
			$watchers{$sv} = $c2->watcher(
				src      => $sv,
				prefetch => $rate,
				rate     => $rate,
				job => sub {
					#shift->requeue(job => shift);
					my ($w,$job) = @_;
					warn "$w ++$job->{id}";
					after {
						$w->requeue(job => $job, cb => sub {
							#warn "
							
						});
					} 0.1;
					#%watchers;
				},
				nomore => sub {
					#%watchers;
				},
			);
		};
		after {
			warn "Starting watcher";
			$neww->();
		} 0.1;
		my $p;$p = periodic {
			$p;
			weaken( my $w = $watchers{$sv} );
			warn "Cleaning watcher ".refcount($w);
			delete $watchers{$sv};
			after {
				if ($w) {
					warn "Not cleaned watcher ".refcount($w);
					print findref $w, 20;
				} else {
					warn "Cleaned";
					$neww->();
				}
			} 1;
			
		} 3;
		return;
		after {
			weaken( my $w = $watchers{$sv} );
			warn "Cleaning watcher ".refcount($w);
			delete $watchers{$sv};
			after {
				if ($w) {
					warn "Not cleaned watcher ".refcount($w);
					print findref $w, 20;
				} else {
					warn "Cleaned";
				}
			} 1;
		} 0.5;
	});
	$cv->recv;
	#warn Dump \%watchers,\%see, $c2;
	exit 0;
}
$0 = "test-pmq - terminal";
$SIG{INT} = sub { kill INT => $server,$loader,$client };

waitpid($server,0) if $server;
waitpid($loader,0) if $loader;
waitpid($client,0);

END {
	kill KILL => -$$ if $server and $client and $loader;
}

__END__


my $start = time;
my ($ins,$upd,$del) = (0)x3;
$| = 1;

$c1->connect(cb {
	warn "Connected: @_";
	#my $w;$w = AnyEvent->timer(interval => $inverval, cb => sub {
	my $w;$w = AnyEvent->timer(after => $inverval, cb => sub {
		undef $w;
		for my $dst (@queues) {
			for my $id (@range) {
				$c1->put(
					dst => $dst,
					id  => $id,
					data => { x => 'x'x10 },
					cb => sub {
						#push @{$history{$id}},"insert: $! ".( @_ > 1 ? "$_[1]" :'').( $taken{$id} ? ' +taken' : '' );
						shift or $!{EEXIST} or warn "create $dst.$id failed: @_";
					},
				);
			}
		}
		return;

		for my $dst (@queues) {
			my $id = $range[ int rand $#range ];
			#warn "Timer $dst $id";
			#$history{$id} ||= [];
			$c1->peek( $dst, $id, cb {
				if (my $j = shift) {
					#warn "Peeked ". Dump $j;
					if (rand > 0.8) {
					#if (rand > 0.2) {
						$del++;
						$c1->delete(
							cb {
								#push @{$history{$id}},"delete: $! ".( @_ > 1 ? "$_[1]" :'').( $taken{$id} ? ' +taken' : '' );
								shift or $!{ENOENT} or warn "Delete failed: @_";
							}
							job => $j,
						);
					} else {
						$upd++;
						$c1->update(
							cb {
								#push @{$history{$id}},"update: $! ".( @_ > 1 ? "$_[1]" :'').( $taken{$id} ? ' +taken' : '' );
								shift or $!{ENOENT} or warn "Update failed: @_ / $!";
							}
							job => $j,
							pri => int rand 100,
							data => { y => 'y'x100 },
						);
					}
				} else {
					$ins++;
					#warn "No job, create";
					$c1->put(
						dst => $dst,
						id  => $id,
						data => { x => 'x'x100 },
						cb => sub {
							#push @{$history{$id}},"insert: $! ".( @_ > 1 ? "$_[1]" :'').( $taken{$id} ? ' +taken' : '' );
							shift or $!{EEXIST} or warn "create $dst.$id failed: @_";
						},
					);
				}
				my $int = time - $start;
				#printf "\r".(" "x40)."\rins: %0.2f/s, upd: %0.2f/s del %0.1f/s  ", $ins/$int, $upd/$int, $del/$int;
			} );
		}
		return;
		#$rc or $self->log->error("Can't update job $exists_job->{src}.$exists_job->{id} in queue: $e");
		#	my ($rc,$e) = 
		#$rc or warn("Can't create job in queue: $e");
	});
});

$cv->recv;
warn Dump \%seen;
